[Lye Zi Foong] iP

lzf834 opened the pull request:
# Duke 🤗

> "The one place for all your tasking needs." - Zi Foong

Duke helps keep track of your tasks, freeing your mind to do the things it needs to do, leave the remembering to Duke. 
It's:

- Text-based
- Easy to use/learn
- Most importantly, it's **FREE**
> "Even my 3-year old knows how to use Duke." John Tan

## Features
### Types of Tasks
- Events
- Todo
- Deadlines
### Cool Stuff
- [X] Managing Tasks with `list`, `mark`, `unmark`, and `delete`
- [X] Deadlines (with both **date** and **time**!)
- [ ] Reminders (coming soon)

## Getting Started

1. Download the jar file from [here](https://github.com/lzf834/ip/releases/tag/A-Jar)
2. Open it up (Double Click)
3. Add in your tasks and events
4. Leave the blues to Duke 😌

If you happen to know how to use Java as well, here's a short snippet of my code:
``` java
public void run() {
        ui.startUp();
        String inputData;
        Scanner scanner = new Scanner(System.in);
        boolean hasEnded = false;

        while (!hasEnded) {
            try {
                inputData = scanner.nextLine();
                hasEnded = parser.takeInput(inputData, taskList);
            } catch (DukeException e) {
                System.out.println(e);
                ui.printSeparator();
            }
            storage.storeTasks(taskList.getTasks());
        }
    }
```

Commit:
Add Gradle support

lzf834 committed:
Level-1 Implemented

lzf834 committed:
Level-2

lzf834 committed:
Level-3 implemeneted

lzf834 committed:
Level-4 implemented

lzf834 committed:
A-TextUiTesting implemented

lzf834 committed:
Level-5 Implemented

lzf834 committed:
Level-6 Implemented

lzf834 committed:
Add Save function, includes file reading and directory creation to keep local list of activities that can be called back

lzf834 committed:
Add Save function, includes file reading and directory creation to keep local list of activities that can be called back

lzf834 committed:
Abstract list method

lzf834 committed:
Merge branch 'branch-Level-7' of https://github.com/lzf834/ip into branch-Level-7

lzf834 committed:
Abstract list, mark, unmark and delete methods from main in duke.java

lzf834 committed:
Merge remote-tracking branch 'origin/master' into branch-Level-7

lzf834 committed:
Add error-handling for wrong inputs for methods mark, unmark and delete

lzf834 committed:
Merge remote-tracking branch 'origin/master' into branch-Level-7

lzf834 committed:
Add date/time handling for the Events and Deadlines

lzf834 committed:
Merge branch 'branch-Level-7'

lzf834 committed:
Merge branch-Level-8 to master

lzf834 committed:
Bug fixing from git merge from Level-8, resolve writing and reading date and time from data.txt file

lzf834 committed:
More OOP

Add Parser class that reads user input
Add Storage class that deals with loading and storing data in data.txt
Add TaskList class, an arrayList that holds all tasks with its own respective methods -> add, delete, print, mark, unmark, and new method "reset" to clear the whole arrayList
Add Ui class that handles print statements

Modified Deadline and Duke.java to accomodate new classes

lzf834 committed:
Organize into packages by placing all classes into duke package

lzf834 committed:
A-JUnit add JUnit tests to test the behaviour of Event, Deadline and Parser class

lzf834 committed:
commit manifest file

lzf834 committed:
A-JavaDoc  Add JavaDoc comments to the code

lzf834 committed:
A-CodingStandard Tweak the code to comply with a given coding standard

lzf834 committed:
Level-9: Implement find function allowing users to find a task by searching for (a) keyword(s)

lzf834 committed:
Merge branch 'branch-A-JavaDoc'

lzf834 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts: resolved
#	src/main/java/duke/Deadline.java
#	src/main/java/duke/Event.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Task.java
#	src/main/java/duke/Todo.java

lzf834 committed:
Merge branch 'branch-Level-9'

# Conflicts resolved:
#	src/main/java/duke/Deadline.java
#	src/main/java/duke/Event.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Task.java
#	src/main/java/duke/Todo.java
#	src/test/java/duke/TaskTest.java

lzf834 committed:
Resolved conflicts in TaskTest.java

lzf834 committed:
Merge branch 'add-gradle-support' of https://github.com/lzf834/ip into branch-A-Gradle

lzf834 committed:
Add and merge gradle to new branch A-gradle, able to run shadowjar test

lzf834 committed:
Merge branch 'branch-A-Gradle'

lzf834 committed:
A-CheckStyle: Add gradle checkstyle and fix styles in files

lzf834 committed:
Merge branch 'branch-A-CheckStyle'

lzf834 committed:
Add png files for GUI

lzf834 committed:
GUI JavaFX tutorial part 1

lzf834 committed:
GUI JavaFx tutorial 1 change font

lzf834 committed:
GUI javaFX tutorial part 2

lzf834 committed:
GUI JavaFX tutorial 3 iteration 1

lzf834 committed:
GUI JavaFX tutorial 3 iteration 2
Update images

lzf834 committed:
GUI JavaFX tutorial 3

JonathanHoshi commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,108 @@
+    public Event(String taskName, String dateTime) throws DukeException {
+        super(taskName);
+
+        dateTime = dateTime.trim();
+        String[] splitString = dateTime.split(" ");
+        if (splitString.length == 1) {
+            try {
+                this.date = LocalDate.parse(splitString[0]);
+            } catch (Exception e) {
+                throw new DukeException("Invalid input into date");
+            }
+        } else if (splitString.length == 2) {
+            try {
+                this.date = LocalDate.parse(splitString[0]);
+                this.time = LocalTime.parse(splitString[1]);
+            } catch (Exception e) {
+                throw new DukeException("Invalid input into date/time");
+            }
+        } else {
+            throw new DukeException("Please input for a date (optional: time)");
+        }
```
Should the logic behind checking for date time inputs be handled in the Parser class where it deals with user input?

JonathanHoshi commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,105 @@
+            }
+        } else if (input.contains("todo") || input.contains("event") || input.contains("deadline")) {
+            //input is a new type of task
+            //identify type of task
+            String[] stringArray = input.split(" ", 2);
+
+            //task has no task detail/name
+            if (stringArray.length < 2) {
+                throw new DukeException("Description of task cannot be empty!");
+            }
+
+            String taskType = stringArray[0];
+            String taskDetails = stringArray[1];
+
+            Task newTask = new Task("");
+
+            if (taskType.equals("todo")) {
+                newTask = new Todo(taskDetails);
+            } else if (taskType.equals("deadline")) {
+                String[] stringSplit = taskDetails.split("/by");
+                if (stringSplit.length < 2) {
+                    throw new DukeException("Description of deadline must include a date/time! Did you miss out a /by?");
+                }
+                String details = stringSplit[0].trim();
+                String dateTime = stringSplit[1].trim();
+                newTask = new Deadline(details,dateTime);
+            } else if (taskType.equals("event")) {
+                String[] splitString = taskDetails.split("/at");
+                if (splitString.length < 2) {
+                    throw new DukeException("Description of event must include a date/time! Did you miss out a /at?");
+                }
+                String details = splitString[0].trim();
+                String dateTime = splitString[1].trim();
+                newTask = new Event(details,dateTime);
+            }
+
+            taskList.addTask(newTask);
+        } else {
+            throw new DukeException("no such task type");
+        }
```
Should the individual code sections of the different commands be extracted? for example 'handleTodo(String input)' and 'handleDelete(String input)'?

JonathanHoshi commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,181 @@
+     * @param input String that contains task type, task details and task date/time if applicable
+     */
+    //Add from String input by user
+    public void addTask(String input) {
```
Any reason why you included this method as the parsing of the input to create a task is already being handled by the Parser class?

JonathanHoshi commented on src/main/java/duke/Todo.java:
```diff
@@ -0,0 +1,46 @@
+     */
+    @Override
+    public void printTask() {
+        System.out.print("[T]");
+        if (this.getDone()) {
+            System.out.println("[X] " + this.getTaskName());
+        } else {
+            System.out.println("[ ] " + this.getTaskName());
+        }
```
Shouldn't you call your toString() method as the code to format the String to the output is already done there?
```suggestion
        System.out.println(toString());
```

JonathanHoshi reviewed (commented):
Overall, good job in structuring your code as it is quite easy to read. I found that in certain parts of your code the nesting was too deep. Consider breaking huge code blocks by separating the logic into segments and splitting them into individual methods to handle each segment. This allows code to be in section to look more organize and neat. Nice work! 👍

lzf834 committed:
GUI JavaFX Tutorial 4 final

lzf834 committed:
Level-10 add GUI for duke

lzf834 committed:
Merge branch 'branch-Level-10'

zhongfu commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * Constructor for Duke class
+     */
+    public Duke(String path, String file_dir) {
```
```suggestion
    public Duke(String path, String fileDir) {
```

variable names should be in camelCase

zhongfu replied:
same for `file_dir` in Storage.java

zhongfu commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,116 @@
+
+    /**
+     * Getter for time
+     * @return time
```
```suggestion
     * 
     * @return time
```

there should be a newline between the description and parameter sections

zhongfu replied:
(same for line 97 of this file, line 90 of Event.java, line 32 of Todo.java)

zhongfu commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,187 @@
+                }
+                String details = stringSplit[0].trim();
+                String dateTime = stringSplit[1].trim();
+                newTask = new Deadline(details,dateTime);
```
```suggestion
                newTask = new Deadline(details, dateTime);
```
commas should be followed by a whitespace

zhongfu replied:
same for line 180, 95, 87

zhongfu reviewed (commented):
I was supposed to only review coding standard violations, but I don't really have much to say in that regard -- your code seems largely compliant, and most of the other mistakes I found would've probably been caught by Checkstyle too. You've already got it set up in Gradle -- you just need to run `gradle checkstyleMain`!

I have some other comments about your code though:
- it feels like you've got some unrelated logic where they shouldn't belong, e.g. printing directly to `System.out` in your Task classes -- possibly a violation of encapsulation
  - `addTask()` vs `addTaskNoPrint()` too -- should the printing of user messages really be the responsibility of your `TaskList` instead of, say, `Ui`? 
  - and many others -- you get the idea
  - now you'd have to change a lot of things across many otherwise unrelated classes if you want to change the message format
- redundant code too: e.g. in your Task classes, you only use `printTime()` in `printDate()`, which in turn is only used in `printTask()`. you *also* reimplement all the logic in `toString()`, which is otherwise identical to `printTask()` except that it returns a `String` instead of printing the same string representation of your Task.
  - same for the date/time parsing logic: should it be repeated in every Task subclass that can take a date/time?
  - `printTask()` and `guiPrintTask()`, and other similar methods too. `guiPrintTask()` should probably be called `toString()` (and override `Object.toString()` too) in light of what it does, which also has the benefit of letting you call `print(task)` instead of `task.printTask()` or `print(guiPrintTask())`, etc
- there's some violation of the abstraction barrier -- should you be leaking implementation details about how your TaskList is implemented outside of the class? what if you no longer want to use an ArrayList?
- your Storage class doesn't seem to escape `~` in user input -- it doesn't result in a fatal error, but users will lose data without warning (e.g. truncated to-dos, or missing events/deadlines)
- your Storage class should also create all parent directories necessary, e.g. using `mkdir()` instead of `mkdirs()`

lzf834 committed:
Edited code to adhere to coding standard

* Add javaDocs for new methods created for GUI-Level-10
* Checkstyle for Java Tests
* JavaDocs for Java Tests inclusive of testing methods

lzf834 committed:
Add assertions (A-Assertions)

* Tasklist - mark, unmark and reset methods, check after each
command that the tasklist/task is in the appropriate state
*Parser and storage - at each stage check that the inputs and their
length is appropriate for parsing

lzf834 committed:
Merge pull request #1 from lzf834/branch-A-Assertions

Add assertions (A-Assertions)

lzf834 committed:
Merge branch 'master' of https://github.com/lzf834/ip

lzf834 committed:
Abstract error messages in Parser.java

lzf834 committed:
Improve Code Quality (A-CodeQuality)

For Parser class, abstract multiple deeply nested checks out into
different methods. e.g. checking for the type of command that the
user input. Abstract date and time checking out as well.

For all tasks classes, removed unnecessary methods that are
repeated such as printdate and printtime, converted to use toString
method

lzf834 committed:
Merge pull request #2 from lzf834/branch-A-CodeQuality

Improve Code Quality (A-CodeQuality)

lzf834 committed:
Merge branch 'master' of https://github.com/lzf834/ip

lzf834 committed:
Add Extension (C-DetectDuplicates)

Duke is now able to detect when a duplicate task is being entered,
if a task that already exists is entered, an error is thrown to the
user, and the task will not be added.

lzf834 committed:
Merge branch 'branch-C-detectDuplicates'

lzf834 committed:
Multiple tasks containing the same details can be added to the list
of tasks, resulting in repeated/duplicated tasks.

This can lead to confusion to the user when he continues to add
tasks that he thinks may not have been added yet to the list.

Let's change the Parser to check whether a task already exists
before adding it.

lzf834 committed:
Text in the windows were very generic and disorganized.

Let's fix that by adding speech bubbles around each text, along
with background colours and a different font to change things up,
making the text more user-friendly.

lzf834 committed:
There is no User Guide, so new users might feel daunted and unsure
of how to use this application.

Let's add a user guide for our new users and guide them through on
Duke's features as well as how to use them.

lzf834 committed:
Fixed some spelling mistakes

lzf834 committed:
Bug Fixing.

1. Fixed bug for CLI print statements. Extra blank line now not
printed anymore

2. Fixed bug for storage. Now data is stored in C drive data folder
instead of a relative data path. JAR file can now store and
retrieve  the data required.

lzf834 committed:
Updated README.md to properly reflect JAR file location.

lzf834 committed:
Abstract methods out to shorten code

lzf834 committed:
Fixed Storage Issues.

Storing in C drive may be an issue for Mac users. Changed storage
location to accommodate for that.

lzf834 committed:
Update Ui.png

lzf834 committed:
no message

lzf834 committed:
Add files via upload

lzf834 committed:
Added more layers of abstraction to fix deep nesting and long
methods

lzf834 committed:
Checkstyle changes

lzf834 committed:
JavaDocs and naming conventions
